### PR TITLE
Governance Model Proposal

### PR DESCRIPTION
**Pull request by blakejohnson**
This introduces a light-weight governance model to the OpenQASM project. Please provide feedback. We aim to merge this within 2-3 weeks, at which point the model becomes "official".

Potential set of initial Working Groups:
* Pulse grammar
* Types and casting (including implicit casts)
* Generics, circuit families, and subroutines

### CODE REVIEW

**Review by jwoehr — COMMENTED** [review]
- Looks good.
- Might state the conditions and procedures to be followed in the event the Governance Body is to be dissolved or absorbed into another body.

**Review by ajavadia — DISMISSED** [review]

File: governance.md
Comment (taalexander) [inline governance.md:45]: Maybe a term of office, and not a term limit? Eg., A TSC member is appointed for two-year term and possibly reelected/affirmed at the end of this term to continue another two-year term. However, there is no limit to the number of terms a TSC member may serve?

File: governance.md
Comment (taalexander) [inline governance.md:43]: 👍

**Review by taalexander — APPROVED** [review]

File: governance.md
Comment (zachschoenfeld33) [inline governance.md:30]:
```suggestion
The TSC responsibility is to balance a desire to allow a voice to the broad OpenQASM community with lightweight processes to enable moving fast and maintaining a consistent philosophy for the language. Thus, for example, this governance document does give the freedom to the TSC to make decisions without calling a meeting and explicitly involving contributors, but the idea is that the TSC should make use of this power judiciously. The TSC has authority to make all decisions regarding OpenQASM (including amendments to this governance document), but should apply more process and seek more community input on constitutional changes than for simple fixes to the spec.
```

File: governance.md
Comment (blakejohnson) [reply governance.md:45]: Yes, a term of office would be better.

File: governance.md
Comment (zachschoenfeld33) [inline governance.md:35]: Is there a centralized location where WG's will be listed? I think it would be beneficial to the community to at least know what issues are currently being explored...

File: governance.md
Comment (zachschoenfeld33) [reply governance.md:45]: I think this suggestion makes sense...

**Review by zachschoenfeld33 — APPROVED** [review]

File: governance.md
Comment (stevenheidel) [reply governance.md:35]: Would be good to add this to the README and also the live spec.

Rust does a really good job at this: https://www.rust-lang.org/governance

**Comment by zachschoenfeld33** [comment]
It is not clear to me when, if ever, the TSC is not needed for a merge. For instance, what about a simple bug fix (like updating the examples) or an ANTLR change? Ideally that wouldn't have to wait for the meeting, especially as PR's may fall out of date with the current spec.

Should the TSC grant authority to certain members who can approve minor fixes? For instance, Hiroshii and I have worked a lot on ANTLR and would be able to help where ANTLR issues arise.

**Comment by jwoehr** [comment]
> It is not clear to me when, if ever, the TSC is not needed for a merge.

"Never needed for a merge" would be the default, don't you think?

Every project has committers. The TSC is a Steering committee, not a code review team. They are _policy_ and _direction_, not day-to-day management.

The committers should be delegated code review as a body. Perhaps 2 positive reviews and no dissents constitute approval to merge. Or whatever the committers as a body may decide under the oversight of the TSC.

Perhaps the committers and the TSC are the same people at first, but that doesn't have to be true.

If a committer or committers is/are irresponsible, then the TSC revokes their status.

This seems rather boilerplate.

File: governance.md
Comment (stevenheidel) [inline governance.md:28]: Two suggestions for this paragraph:
1. I would like to insert some language that the TSC aims to build consensus among contributors prior to voting in most cases. Voting to resolve disagreements is ideally expected to happen less often than voting to accept the consensus. There is no way to write this into the structure but it's a goal we should strive towards.
2. Allow the TSC to invite guests to the TSC meetings, even if they are not contributors yet.

File: governance.md
Comment (stevenheidel) [inline governance.md:15]: Clarify that contributors are non-voting.

File: governance.md
Comment (stevenheidel) [inline governance.md:32]:
```suggestion
The TSC may have a **Secretary** who organizes the meeting (e.g. prepares and distributes the agenda and collects a set of issues to be considered). The TSC **Secretary** will be a **Contributor** appointed by vote of the TSC.
```

"non-voting Contributor" is confusing since all contributors are non-voting.

**Comment by stevenheidel** [comment]
Re: merges - I agree with @jwoehr. If there are multiple contributors in support of a change and no one opposed then there is no need to involve the TSC. I expect doc changes, ANTLR fixes, etc to fall into this bucket.

We are using git so as long as the main branch is protected then even in the worst case we can revert a commit back into a PR. Repeated irresponsible behavior would be grounds for revoking contributor status.

We need to add language on how the contributor status would be revoked.

> Contributors are selected by nomination of at least 2 current TSC Members.

Suggest adding something like "Contributor status may be revoked at the discretion of the TSC".

File: governance.md
Comment (blakejohnson) [reply governance.md:35]: Wow, that is nice. Having something as fancy would clearly take infrastructure we don't have yet, but listing WGs in the README would be a good start.

**Comment by blakejohnson** [comment]
@jwoehr: Adding to what @stevenheidel said, I agree that there will be circumstances where the TSC does not need to be involved. I would very much like to reach the point where many issues and PRs can be resolved by Contributors. However, what you outlined will mostly apply to code, tooling, and bug fixes. Significant changes to the language itself are in a different category that absolutely should hit the pause button to allow for centralized approval. For example, a major language change or addition like #184 will *require* explicit TSC approval. Does that make sense?

**Comment by jwoehr** [comment]
Abostively, my friend ... as a CEO once said in a heated meeting, "I think we're in violent agreement here." :)

> @jwoehr: Adding to what @stevenheidel said, I agree that there will be circumstances where the TSC does not need to be involved. I would very much like to reach the point where many issues and PRs can be resolved by Contributors. However, what you outlined will mostly apply to code, tooling, and bug fixes. Significant changes to the language itself are in a different category that absolutely should hit the pause button to allow for centralized approval. For example, a major language change or addition like #184 will _require_ explicit TSC approval. Does that make sense?

**Comment by zachschoenfeld33** [comment]
> @jwoehr: Adding to what @stevenheidel said, I agree that there will be circumstances where the TSC does not need to be involved. I would very much like to reach the point where many issues and PRs can be resolved by Contributors. However, what you outlined will mostly apply to code, tooling, and bug fixes. Significant changes to the language itself are in a different category that absolutely should hit the pause button to allow for centralized approval. For example, a major language change or addition like #184 will _require_ explicit TSC approval. Does that make sense?

I agree w/ this as well. I think some language clarifying this could be helpful, just so it's in the document, in addition to grounds for revoking contributor status as noted by @stevenheidel. We might also say any commit/merge in this category can be undone at any time by the TSC (this may be obvious and not necessary to state).

The other reason I asked is because "merge" power is heavily restricted, only a few people have it as far as I know, rather than any contributor...

**Comment by blakejohnson** [comment]
> The other reason I asked is because "merge" power is heavily restricted, only a few people have it as far as I know, rather than any contributor...

Once this is merged, I would be happy to extend "merge" power to all Contributors.

**Comment by zachschoenfeld33** [comment]
When this is approved, we can merge in https://github.com/Qiskit/openqasm/pull/221, which is a list defining the current and former OpenQASM working groups.

**Comment by zachschoenfeld33** [comment]
> Seems there are three levels:
> 
> 1. TSC describes the language and governs the project.
> 2. COMMITTERS are trusted persons empowered to merge after whatever community review and consultation (if needed) with the TSC has, in their practice, been performed.
> 3. CONTRIBUTORS are anybody who forks and PR's and has a PR accepted.
> 
> Probably you end up with a program where over time successful and trusted CONTRIBUTORS are promoted to COMMITTERS when a COMMITTER leaves the project, or the project expands, or whatever.
> 
> This is approximately how the Apache NetBeans project works.

I think this makes sense, depending on how many layers we want. Either way could work (with or without the "committer" layer)

**Comment by jwoehr** [comment]
> I  think this makes sense, depending on how many layers we want. Either way could work (with or without the "committer" layer)

Assuming that the Qiskit project continues to evolve, I infer that the senior persons on the TSC will welcome an administrative layer to handle routine operations such as vetting code and doc PRs from outside contributors. The "committer" level can be trusted to carry out the guidance of the TSC and examining PRs for conformance to the model.

File: governance.md
Comment (blakejohnson) [inline governance.md:7]: To clarify what it means to be a "Qiskit project":
```suggestion
This governance structure is scoped to just OpenQASM itself and not Qiskit as a whole. The scope of OpenQASM includes the language specification as well as reference implementations of parsers or other low-level infrastructure to facilitate use in other projects. OpenQASM inherits the contribution guidelines and community standards that are common to all projects within the Qiskit GitHub organization.
```

**Comment by blakejohnson** [comment]
The bar we are setting for a Contributor is higher than merely having a PR accepted. It is demonstration of sustained involvement. So, I would like not to introduce an extra layer but simply give Contributors merge privs.

File: governance.md
Comment (blakejohnson) [reply governance.md:28]: It effectively postpones the vote until the meeting. TSC members might then choose to change their votes.

File: governance.md
Comment (blakejohnson) [reply governance.md:32]: The secretary cannot break ties, but they can participate in the discussion.

File: governance.md
Comment (blakejohnson) [reply governance.md:45]: If we do terms, I suggest staggering the initial terms so that the election process isn't choosing an entire TSC at once.

File: governance.md
Comment (blakejohnson) [reply governance.md:15]: Addressed in 70baf4cbd88f9705d78206d061f3e16b6614b8a7

File: governance.md
Comment (blakejohnson) [reply governance.md:28]: Addressed in 70baf4cbd88f9705d78206d061f3e16b6614b8a7

File: governance.md
Comment (blakejohnson) [reply governance.md:32]: Addressed in 70baf4cbd88f9705d78206d061f3e16b6614b8a7

**Review by ajavadia — DISMISSED** [review]

File: governance.md
Comment (levbishop) [inline governance.md:36]:
```suggestion
**TSC Members** will serve a 2-year term, with the exception of the founding TSC which will have staggered terms to avoid an election of an entire TSC all at once. New **TSC Members** will be selected by nomination of the TSC, followed by a vote which is open to all **Contributors**. Outgoing **TSC Members** participate in the nomination process which selects their successors. Note that **TSC Members** may be re-elected to serve additional 2-year terms. The same process will be used in the event that a **TSC Member** resigns prior to the end of their term.
```

**Review by levbishop — APPROVED** [review]
Noted a typo otherwise ok for me

**Review by taalexander — APPROVED** [review]

**Review by levbishop — APPROVED** [review]
Ship it 🛳

**Review by stevenheidel — APPROVED** [review]
Thank you @blakejohnson et al for your work on formalizing this

**Pull request merged**
